manifests/instance/dn/add.pp

```diff
@@ -0,0 +1,167 @@
+# populated.
+#
+# This works by comparing a generated LDIF with the LDIF formatted contents of
+# the target instance.
+#
+# If passing a full LDIF - DO NOT WRAP LINES
+#
+# @option name [String]
+#   A globally unique description of the desired configuration setting
+#
+# @param instance_name
+#   The instance name as passed to `ds389::instance`
+#
+# @param dn
+#   The DN to be created
+#
+# @param objectclass
+#   objectClasses to which the DN belongs
+#
+# @param attrs
+#   Attributes that you wish to set at the time of creation
+#
+# @param content
+#   The full content of the LDIF
+#
+#   * This may only contain *one* entry
+#   * All other parameters will be ignored
+#   * DO NOT add 'changetype: add'
+#
+# @param root_dn
+#   A DN with administrative rights to the directory
+#
+#   * Will be determined automatically if not set
+#
+# @param root_pw_file
+#   A file containing the password for use with ``$root_dn``
+#
+#   * Defaults to `$ds389::config_dir/<usual pw file>`
+#
+# @param port
+#   The port to which to connect
+#
+#   * Has no effect if LDAPI is enabled on the instance
+#   * Will be determined automatically if not set
+#
+# @param force_ldapi
+#   Force the system to use the LDAPI interface
+#
+#   * Generally only useful during bootstrapping
+#
+# @param restart_instance
+#   Whether or not to restart the directory server after applying this item
+#
+define ds389::instance::dn::add (
+  Simplib::Systemd::ServiceName         $instance_name,
+  Optional[Pattern['^\S+=.+']]          $dn               = undef,
+  Optional[Array[String[1],1]]          $objectclass      = undef,
+  Optional[Hash[String[1],String[1],1]] $attrs            = undef,
+  Optional[String[3]]                   $content          = undef,
+  Optional[String[2]]                   $root_dn          = undef,
+  Optional[Stdlib::Absolutepath]        $root_pw_file     = undef,
+  Optional[Simplib::Host]               $host             = undef,
+  Optional[Simplib::Port]               $port             = undef,
+  Boolean                               $force_ldapi      = false,
+  Boolean                               $restart_instance = false
+) {
+  $_instance_name = split($instance_name, /^(dirsrv@)?slapd-/)[-1]
+  $_filesafe_dn = regsubst($dn, '[\\\\/:~\n\s\+\*\(\)@]', '__', 'G')
+
+  if !$dn and !content {
+    fail('You must specify either $dn or $content')
+  }
+  if $dn and !($objectclass or $attrs) {
+    fail('You must specify $objectclass and $attrs with $dn')
+  }
+
+  $_known_instances = pick($facts['ds389__instances'], {})
+
+  $_root_dn = pick($root_dn, pick($_known_instances.dig($_instance_name, 'rootdn'), false))
+  unless $_root_dn {
+    fail("You must specify an \$root_dn for '${title}'")
+  }
+
+  $_root_pw_file = pick($root_pw_file, "/usr/share/puppet_ds389_config/${_instance_name}_ds_pw.txt")
+
+  if $force_ldapi or $_known_instances.dig($_instance_name, 'ldapilisten') {
```

Comment:
this code block looks the same as code in `ds389::instance::attr::set`.  Could it be encapsulated in a function?

Reply:
I can pop it into a class. I'm trying to avoid functions for things like this because it's "just a manifest" and doesn't seem to warrant function-level complexity.

Reply:
Actually, I'm not going to do this. I actually thought about it earlier but it saves almost no space and makes everything more difficult to understand. Technically, all of this stuff should go into a native type at some point.